From b11f50d6186aba5c2493114d725a82634523ec10 Mon Sep 17 00:00:00 2001
From: Peter Martienssen <peter.martienssen@liquify-consulting.de>
Date: Fri, 5 Jun 2026 11:53:01 +0200
Subject: [PATCH] media: imx8mp-csi: clear GASKET data type before updating
 format

The CSI gasket data type field was updated using a bitwise OR
operation without clearing the previous value first.

Switching between RAW10 (0x2b) and RAW12 (0x2c) therefore left
stale bits in the field and resulted in an invalid data type
value of 0x2f being programmed into GPR_GASKET_0_CTRL.

This prevented the ISI from receiving valid image data after
a pixel format change until the system was rebooted.

Fix this by clearing the DATA_TYPE and DUAL_COMP_ENABLE bits
before programming the new gasket configuration.
---
 drivers/staging/media/imx/imx8-mipi-csi2-sam.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/staging/media/imx/imx8-mipi-csi2-sam.c b/drivers/staging/media/imx/imx8-mipi-csi2-sam.c
index 5e850497fd13..11883a9691f6 100644
--- a/drivers/staging/media/imx/imx8-mipi-csi2-sam.c
+++ b/drivers/staging/media/imx/imx8-mipi-csi2-sam.c
@@ -1045,6 +1045,7 @@ static void disp_mix_gasket_config(struct csi_state *state)
 	}
 
 	regmap_read(gasket, DISP_MIX_GASKET_0_CTRL, &val);
+	val &= ~(GASKET_0_CTRL_DATA_TYPE_MASK | GASKET_0_CTRL_DUAL_COMP_ENABLE);
 	if (fmt_val == GASKET_0_CTRL_DATA_TYPE_YUV422_8)
 		val |= GASKET_0_CTRL_DUAL_COMP_ENABLE;
 	val |= GASKET_0_CTRL_DATA_TYPE(fmt_val);
-- 
2.34.1

